### PR TITLE
fix ctx->msg for cold events

### LINKED ISSUE #2902: VGC and VMOD entrypoints.
This ticket replaces #2897 and #2898 

I have surveyed all the ways we call into VGC and VMOD code, and it is not very … consistent.

We have calls in a CLI context, which are allowed to fail, and expected to fail if something is not right, most notably VCL_EVENT_LOAD.

Here a ctx->msg is provided for communicating an error message back to the CLI.

We have calls in a CLI(-like) context which are not allowed to fail. VCL_EVENT_DISCARD, _COLD, _WARM, and no ctx->msg is provided, because failure is not an option.

(There are further nuances, VCL_EVENT_LOAD and _DISCARD sets ctx->method to VCL_RET_INIT and _FINI respectively, because VCL code is called, the other events _COLD & _WARM leave method at zero.)

We also have calls in CLI context to simply gather information, currently VDI->list(), which are not allowed to fail and where ctx->msg is the output buffer.

In none of these CLI(-like) contexts does PRIV_TASK make any sense, and there is never a ctx->vsl.

Then we have all the traffic-entries to execute compiled VCL code, which is where PRIV_TASK makes sense, ctx->msg does not, and ctx->vsl is always present.

I will 100% admit that it can be convenient for VMOD code to not have to bother about which kind of context it is operating in, and just grap PRIV_TASK or VRT_fail() and expect The Right Thing to happen, but I think it is the wrong way to go, because it adds code and overhead for very little saving.

I think my proposal going forward is:

In CLI(-like) contexts, there will always be a ctx->msg, and unless ctx->handling is VCL_RET_OK on return, ctx->msg contains a diagnostic.

This makes it possible to call VRT_fail() in all VGC and VMOD code.

The code calling into VGC/VMOD code may still be intolerant of failure, in which case it should report ctx->msg and panic - as we do today.

Information gathering code (like VDI->list()) can either panic on failure or substitute a "canned" response instead of ctx->msg, as makes most sense.

If a canned response is used, ctx->msg should be emitted somewhere (probably VSL(SLT_Error) ?)

cli_priv_task will be removed, and VGC/VMOD code is responsible for testing ctx->vsl before using VMOD_PRIV related functionality.

### PR DESCRIPTION
**Pull request by nigoroll**
This text is a slightly edited version of the commit message from the third commit, as the other two are only preparations.

This PR adds the final bit to fix #2902: As discussed in the ticket, there should be no `VRT_CTX` `msg` vsb present for `COLD` events, yet existing code did provide it in some cases.

Comments on the most relevant changes to `cache_vcl.c` in order of appearance:

Existing code kept the CLI `VRT_CTX` around for longer than it strictly needed to just because access to the error message in `ctx->msg` was still required.

In order to be able to simplify management of the CLI `VRT_CTX`, we change `VCL_Rel_CliCtx()` to return the `ctx->msg` IFF it was requested by a _true_ argument to `VCL_Get_CliCtx()`

Regarding #2902, the main issue with existing code was that the decision on whether or not to request a `ctx->msg` was separated from where we could actually make that decision: It is (only) in
`vcl_send_event()` where we know which event is going to be issued, which, in turn, determines if we are going to need a `ctx->msg`.

Thus, we move the `VRT_CTX` management for all of the CLI operations into one place in `vcl_send_event()` and change the infrastructure around it to handle the vcl and the error message vsb instead of a `VRT_CTX`.

This allows us to (finally) ensure that `VCL_EVENT_COLD` does never have a `ctx->msg` error vsb, asserted by `AZ(havemsg)` in the switch/case statement in `vcl_send_event()`.

In `vcl_send_event()`, we also assert that if a vcl subroutine was called in the case of a LOAD or DISCARD event, only `vcl_init {}` for a LOAD event may return anything but OK.

`vcl_set_state()` contains the two use cases of `vcl_send_event()`: When we know that an event may not fail, we also assert that there is no msg vsb returned, and if it could, we keep it, as it is still required by its callers.

`vcl_cancel_load()` shows how the new infrastructure simplifies the code: Rather than having to re-setup the `VRT_CTX` for the DISCARD event, we simply output the error and have `vcl_send_event()` set up the proper handling for that event.

Basically the same pattern repeats in `vcl_load()`, `VCL_Poll()` and `vcl_cli_state()`: In these functions, we only look after error handling and output.

Fixes #2902

### CODE REVIEW

**Comment by nigoroll** [comment]
explained during bugwash, no objections, still needs code review

File: bin/varnishd/cache/cache_vcl.c [hunk 1]
```diff
@@ -160,31 +160,43 @@ VCL_Rel_CliCtx(struct vrt_ctx **ctx)
...
+	case VCL_EVENT_DISCARD:
+		ctx->method = VCL_MET_FINI;
+		/* FALLTHROUGH */
+	case VCL_EVENT_COLD:
+		// XXX AZ(ctx->msg);
+		break;
+	default:
+		WRONG("vcl_event");
+	}
+
```
[the remark this answers is not in the document]
Comment (nigoroll) [reply bin/varnishd/cache/cache_vcl.c]: we do not have a method for warm/cold.
I actually considered adding `AZ(method)` in the switch/case, but found that did not contribute to clarity much. If you disagree, we can change that

File: bin/varnishd/cache/cache_vcl.c [hunk 2]
```diff
@@ -140,62 +140,91 @@ VCL_Get_CliCtx(int msg)
...
+	struct vsb *r = NULL;
 
 	ASSERT_CLI();
 	assert(*ctx == &ctx_cli);
 	AN((*ctx)->handling);
-	if (ctx_cli.msg)
-		VSB_destroy(&ctx_cli.msg);
+	if (ctx_cli.msg) {
+		TAKE_OBJ_NOTNULL(r, &ctx_cli.msg, VSB_MAGIC);
+		VSB_finish(r);
```
[the remark this answers is not in the document]
Comment (nigoroll) [reply bin/varnishd/cache/cache_vcl.c]: yes, thank you

File: bin/varnishd/cache/cache_vcl.c [hunk 3]
```diff
@@ -499,39 +528,37 @@ VCL_TestLoad(const char *fn)
...
+	msg = VSB_new_auto();
+
+	VSB_printf(msg, "VCL %s is waiting for:", vcl->loaded_name);
 	Lck_Lock(&vcl_mtx);
-	VTAILQ_FOREACH(ref, &ctx->vcl->ref_list, list)
-		VSB_printf(ctx->msg, "\n\t- %s", ref->desc);
+	VTAILQ_FOREACH(ref, &vcl->ref_list, list)
+		VSB_printf(msg, "\n\t- %s", ref->desc);
 	Lck_Unlock(&vcl_mtx);
+	VSB_finish(msg);
```
[the remark this answers is not in the document]
Comment (nigoroll) [reply bin/varnishd/cache/cache_vcl.c]: same

File: bin/varnishd/cache/cache_vcl.c [hunk 4]
```diff
@@ -499,39 +528,37 @@ VCL_TestLoad(const char *fn)
 
 /*--------------------------------------------------------------------*/
 
-static void
-vcl_print_refs(VRT_CTX)
+static struct vsb *
+vcl_print_refs(struct vcl *vcl)
```
[the remark this answers is not in the document]
Comment (nigoroll) [reply bin/varnishd/cache/cache_vcl.c]: right

**Comment by nigoroll** [comment]
@bsdphk thank you for the review. I have pushed commits addressing your feedback except for the first comment.
